File: src/services/mcp-service.ts
import { AuthorizationError, NotFoundError } from '../domain/errors.js';
import {
  CreateMcpListingInput,
  McpListing,
  SearchMcpListingsInput,
} from '../domain/models.js';
import { validateCreateListingInput, validateSearchInput } from '../domain/validators.js';
import { Clock } from '../infrastructure/clock.js';
import { createId } from '../infrastructure/ids.js';
import {
  CreateListingRecord,
  McpListingRepository,
} from '../repositories/mcp-listing-repository.js';

export interface McpService {
  createListing(ownerId: string, input: CreateMcpListingInput): McpListing;
  deleteListing(requestUserId: string, listingId: string): void;
  searchPublic(input: SearchMcpListingsInput): McpListing[];
  listOwnedByUser(userId: string): McpListing[];
}

export function createMcpService(deps: {
  listings: McpListingRepository;
  clock: Clock;
}): McpService {
  return {
    createListing(ownerId, input) {
      const validated = validateCreateListingInput(input);
      const createdAt = deps.clock.nowIso();

      const record: CreateListingRecord = {
        id: createId(),
        ownerId,
        title: validated.title,
        summary: validated.summary,
        version: validated.version,
        transport: validated.transport,
        authType: validated.authType,
        sourceType: validated.sourceType,
        endpointUrl: validated.endpointUrl ?? null,
        packageName: validated.packageName ?? null,
        packageRegistry: validated.packageRegistry ?? null,
        sourceUrl: validated.sourceUrl ?? null,
        homepageUrl: validated.homepageUrl ?? null,
        tagsJson: JSON.stringify(validated.tags ?? []),
        taskTypesJson: JSON.stringify(validated.taskTypes ?? []),
        isDeleted: 0,
        createdAt,
        deletedAt: null,
      };

      deps.listings.create(record);
      const created = deps.listings.findById(record.id);
      if (!created) {
        throw new NotFoundError('创建后的 MCP 记录未找到');
      }

      return created;
    },
    deleteListing(requestUserId, listingId) {
      const listing = deps.listings.findById(listingId);
      if (!listing || listing.isDeleted) {
        throw new NotFoundError('未找到对应的 MCP');
      }

      if (listing.ownerId !== requestUserId) {
        throw new AuthorizationError('只能删除自己上传的 MCP');
      }

      deps.listings.softDelete(listingId, deps.clock.nowIso());
    },
    searchPublic(input) {
      const validated = validateSearchInput(input);
      return deps.listings.searchPublic(validated);
    },
    listOwnedByUser(userId) {
      return deps.listings.findOwnedByUser(userId);
    },
  };
}

File: src/services/mcp_service.py
from src.domain.errors import AuthorizationError, NotFoundError
from src.domain.models import CreateMcpListingInput, McpListing, SearchMcpListingsInput
from src.domain.validators import validate_create_listing_input, validate_search_input
from src.infrastructure.clock import now_iso
from src.infrastructure.ids import create_id
from src.repositories.mcp_listing_repository import McpListingRepository


class McpService:
    def __init__(self, listings: McpListingRepository) -> None:
        self._listings = listings

    def create_listing(self, owner_id: str, input_data: CreateMcpListingInput) -> McpListing:
        validated = validate_create_listing_input(input_data)
        listing = McpListing(
            id=create_id(),
            owner_id=owner_id,
            title=validated.title,
            summary=validated.summary,
            version=validated.version,
            transport=validated.transport,
            auth_type=validated.auth_type,
            source_type=validated.source_type,
            endpoint_url=validated.endpoint_url,
            package_name=validated.package_name,
            package_registry=validated.package_registry,
            source_url=validated.source_url,
            homepage_url=validated.homepage_url,
            tags=validated.tags,
            task_types=validated.task_types,
            is_deleted=False,
            created_at=now_iso(),
            deleted_at=None,
        )
        self._listings.create(listing)
        created = self._listings.find_by_id(listing.id)
        if not created:
            raise NotFoundError("创建后的 MCP 记录未找到")
        return created

    def delete_listing(self, request_user_id: str, listing_id: str) -> None:
        listing = self._listings.find_by_id(listing_id)
        if not listing or listing.is_deleted:
            raise NotFoundError("未找到对应的 MCP")

        if listing.owner_id != request_user_id:
            raise AuthorizationError("只能删除自己上传的 MCP")

        self._listings.soft_delete(listing_id, now_iso())

    def search_public(self, input_data: SearchMcpListingsInput) -> list[McpListing]:
        return self._listings.search_public(validate_search_input(input_data))

    def list_owned_by_user(self, user_id: str) -> list[McpListing]:
        return self._listings.find_owned_by_user(user_id)
